---
title: "Hisat2_Stringtie2_Ballgown"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# For RNA-seq reads, use "--dta/--downstream-transcriptome-assembly" 
# the command is similar to 'bowtie2'

$ hisat2 -x genome -1 read1.fq.gz -2 read2.fq.gz -S Sample.sam -p 8 --dta  2>&1 | tee $l.stat 

$ samtools view -bS Sample.sam | samtools sort -@ 8 - Sample.sorted


# The bam files can be fed into other pipelines e.g DESeq, edgeR, and etc
# transcriptome assembly using 'stringtie'

$ stringtie -p 8 -G genes.gtf -o Sample.gtf –l Sample.sorted.bam

$ stringtie --merge -p 8 -G genes.gtf -o merged.gtf mergelist.txt

# expression level estimation
$ stringtie –e –B -p 8 -G merged.gtf -o Sample.gtf Sample.sorted.bam

# count data can be extracted from 'stringtie' output using 'prepDE.py', 
# and you can use them for DESeq/edgeR analysis


# Run the differential expression analysis protocol using 'Ballgown'
