---
title: "Report of RNA-seq Analysis of Project IGC-BB-19103"
author: | 
 |
 | Yuqi Zhao, PhD; Xiwei Wu, MD, PhD
 |
 | Integrative Genomics Core, Beckman Research Institute of City of Hope
 | 
 |   
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    keep_md: yes
  pdf_document:
    toc: yes
---
  
The RNA-seq data analysis contains:  
1) Quality Control (fastp/multiqc),  
2) Reads Alignments (hisat2),  
3) Transcriptome Assembly (stringtie),  
4) Data Normalization (ballgown/DESeq2),
5) Differentially Expressed Analysis (ballgown/DESeq2),  
6) Pathway Enrichment (enrichR/GSEA).  


```{r setup, include=FALSE,echo=FALSE,message = FALSE,warning = FALSE,error=TRUE}
#  
# if (!requireNamespace("BiocManager", quietly = TRUE))  
#   install.packages("BiocManager")  
# BiocManager::install(c("factoextra","ballgown","dplyr","pheatmap","cowplot","ggplot2","qusage","msigdbr","fgsea","kableExtra","enrichR","EnhancedVolcano","magrittr","ggpubr","DESeq2","rtracklayer"))  
#Change Work Directory and install all the required packages   
setwd("/Volumes/YZWD/Projects_COH/RNA_seq/IGC-BB-19103/mm10/DESeq2/")
suppressPackageStartupMessages(library(factoextra))
suppressPackageStartupMessages(library(ballgown))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(qusage))
suppressPackageStartupMessages(library(msigdbr))
suppressPackageStartupMessages(library(fgsea))
suppressPackageStartupMessages(library(kableExtra))
suppressPackageStartupMessages(library(enrichR))
suppressPackageStartupMessages(library(EnhancedVolcano))
suppressPackageStartupMessages(library(magrittr))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(rtracklayer))
# download.file("http://www.gsea-msigdb.org/gsea/msigdb/download_file.jsp?filePath=/msigdb/release/7.4/c2.cp.v7.4.symbols.gmt",destfile = "c2.cp.v7.4.symbols.gmt")
```
## Quality Control  

Quality control and preprocessing of FASTQ files are done using [fastp](https://github.com/OpenGene/fastp). The separate reports of quality control are pooled using [MultiQC](https://multiqc.info/).  


```{r Quality Control,fig.height = 8,fig.width = 12,include = TRUE,echo=FALSE,message = FALSE,warning = FALSE,error=TRUE}

#Citation:  
#[Shifu Chen, Yanqing Zhou, Yaru Chen, Jia Gu; fastp: an ultra-fast all-in-one FASTQ preprocessor, Bioinformatics, Volume 34, Issue 17, 1 September 2018, Pages i884–i890](https://academic.oup.com/bioinformatics/article/34/17/i884/5093234).  

# Install fastp
# note: the fastp version in bioconda may be not the latest
#conda install -c bioconda fastp

#!/bin/bash
#SBATCH --array=1-13
#SBATCH -N 1
#SBATCH -n 2
#SBATCH --mem 50G
#SBATCH -t 0-72:00:00
#SBATCH –p compute
#SBATCH -q batch

# cd $SLURM_SUBMIT_DIR
# source /home/yuqzhao/anaconda3/etc/profile.d/conda.sh
# conda activate Anacondaent3
# sample=$(cat samples.txt|head -n $SLURM_ARRAY_TASK_ID| tail -1)
# mkdir -p ../Trimming/
# fastp -h "$sample".html -j "$sample".json -i "$sample"_R1.fastq.gz -I "$sample"_R2.fastq.gz -o ../Trimming/"$sample"_R1.fastq.gz -O ../Trimming/"$sample"_R2.fastq.gz
RNA_seq_samples<-read.table("RNA_seq_samples.txt",sep = "\t",header = TRUE)
kbl(RNA_seq_samples,caption = "Table 1. RNA-seq samples in the study") %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"),font_size = 11)

```

![Barplot of reads filtering in fastp](/Volumes/YZWD/Projects_COH/RNA_seq/IGC-BB-19103/mm10/DESeq2/fastp_filtered_reads_plot.png)  
The RNA-seq data is of high quality, with over 97% of the raw reads passing the quality filtering in fastp. The details of quality control can be found in Supplementary multiqc_report.html.

## Reads Alignments   
**HISAT2** is adopted for mapping next-generation sequencing reads to a reference genome.  

```{r Reads Alignments,fig.height = 8,fig.width = 12,include = TRUE,echo=FALSE,message = FALSE,warning = FALSE,error=TRUE}

# Citations:  
# [Pertea, M., Kim, D., Pertea, G. et al. Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown. Nat Protoc 11, 1650–1667 (2016).] (https://doi.org/10.1038/nprot.2016.095)  

# The basic steps include:
# 1) Align clean reads to referecne genome  
# <span style="color:blue">hisat2 -x referecne_genome -1 Read1 -2 Read2 -S Output.sam</span>.  
# 2) sorted the alignments in "genome" order  
# <span style="color:blue">samtools sort -o Output.sorted.bam Output.sam</span>.  
# 3) Remove duplicate reads  
# <span style="color:blue">picard MarkDuplicates I=Output.sorted.bam O=Output.sorted.nodup.bam M=Output.metric.txt REMOVE_DUPLICATES=true</span>.  
# 4) Create index file  
# <span style="color:blue">samtools index Output.sorted.nodup.bam</span>.    

# Scripts:
  
#!/bin/bash
#SBATCH --array=1-13%5
#SBATCH -N 1
#SBATCH -n 4
#SBATCH --mem 200G
#SBATCH -t 0-24:00:00
#SBATCH –p compute
#SBATCH -q batch
#cd $SLURM_SUBMIT_DIR
#sample=$(cat samples.txt|head -n $SLURM_ARRAY_TASK_ID| tail -1)
#source ~/anaconda3/etc/profile.d/conda.sh
#conda activate Anacondaent3
#module load HISAT2/2.1.0-foss-2017a
#module load picard/2.21.1
#module load samtools/1.6
#fastq-dump.2.9.2 --split-files $sample
#GENOME_Index="/home/yuqzhao/GENCODE_Human/Mouse/GRCm39"
#
#fastp -i "$sample".fastq -o "$sample"_fastp.fastq -f 5 -t 6
#Read=./"$sample"_fastp.fastq
#Read1=/home/yuqzhao/Yuqi_Projects/RNA_seq/IGC-BB-19103/Trimming/"$sample"_R1.fastq.gz
#Read2=/home/yuqzhao/Yuqi_Projects/RNA_seq/IGC-BB-19103/Trimming/"$sample"_R2.fastq.gz
#hisat2 -p 4 -x $GENOME_Index -U $Read -S "$sample".sam
#hisat2 -p 4 -x $GENOME_Index -1 $Read1 -2 $Read2 -S "$sample".sam
#samtools sort -@ 4 -o "$sample".sorted.bam "$sample".sam

#java -jar /opt/picard/2.21.1/picard.jar MarkDuplicates I="$sample".sorted.bam O="$sample".sorted.nodup.bam M="$sample".
#metric.txt REMOVE_DUPLICATES=true
#samtools index "$sample".sorted.nodup.bam

#rm "$sample".sam
#rm "$sample".sorted.bam



# The Reads Alignments outputs includes:  
# 1) bam files without duplicates (*.sorted.nodup.bam),  
#2) index files (*.sorted.nodup.bam.bai). 

Alignment_Rates<-read.table("Alignment_Rates.txt",sep = "\t",header = TRUE)
ggplot(Alignment_Rates, aes(fill=Treatments, y=Alignment_Rates, x=Sample_Names)) + 
  geom_bar(position="dodge", stat="identity") + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
```

The overall read alignment rates were above 96 % for all libraries.  

The Integrative Genomics Viewer (IGV) plot shows the RNA-seq alignments across all the samples. 
![IGV plot of reads alignments in gene Errifi1.](/Volumes/YZWD/Projects_COH/RNA_seq/IGC-BB-19103/mm10/DESeq2/Alignment_Coverage_igv.png)

![The knockout efficiency tend to be higher at exon 3 and 4.](/Volumes/YZWD/Projects_COH/RNA_seq/IGC-BB-19103/mm10/DESeq2/igv_snapshot_exon_2_3.png)
  
  
## Transcriptome Assembly  
**StringTie** is adopted for fast and highly efficient assembly of RNA-Seq alignments into potential transcripts.  
**Ballgown** is applied to facilitate flexible differential expression analysis of RNA-seq data.  
The normalization in **DESeq2** is included in the RNA-seq analysis.  
```{r Transcriptome Assembly,fig.height = 8,fig.width = 12,include = TRUE,echo=FALSE,message = FALSE,warning = FALSE,error=TRUE}
# Citations:  
# [Pertea, M., Kim, D., Pertea, G. et al. Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown. Nat Protoc 11, 1650–1667 (2016).] (https://doi.org/10.1038/nprot.2016.095)

#!/bin/bash
#SBATCH --array=1-13%5
#SBATCH -N 1
#SBATCH -n 2
#SBATCH --mem 20G
#SBATCH -t 0-24:00:00
#SBATCH –p compute
#SBATCH -q batch
#cd $SLURM_SUBMIT_DIR
#module load stringtie
#mkdir -p ballgown
#sample=$(cat samples.txt|head -n $SLURM_ARRAY_TASK_ID| tail -1)
#stringtie -e -B -p 2 -G /net/isi-dcnl/ifs/user_data/Seq/Yuqi_Projects/CCLE/RNA_seq/gencode.v36.annotation.gtf -o ballgown/$sample/"$sample".gtf "$sample".sorted.nodup.bam

# Main Outputs
# 1) Stringtie's main output is a GTF file containing the assembled transcripts  
# 2) Gene abundances in tab-delimited format  
# 3) Fully covered transcripts that match the reference annotation, in GTF format  
# 4) Files (tables) required as input to Ballgown, which uses them to estimate differential expression  
# 5) In merge mode, a merged GTF file from a set of GTF files  
# 
#  
# **prepDE.py** derives hypothetical read counts for each transcript from the coverage values estimated by **StringTie** for each transcript, by using this simple formula: reads_per_transcript = coverage * transcript_len / read_len  
# The format of gtf_files.txt should be:
# Sample1 Path_to_sample1_gtf  
# Sample2 Path_to_sample2_gtf  
# ...
#   
# <span style="color:blue">python prepDE.py3 -i gtf_files.txt</span>.  
#   
# generates two CSV files containing the count matrices for genes and transcripts.  
# 
mm10_gtf_df<-data.frame(import("mm10.ncbiRefSeq.gtf"))
transcript2genes<-mm10_gtf_df %>% filter(type=="transcript") %>% dplyr::select(gene_name,transcript_id)
countData_Gene <- as.matrix(read.csv("gene_count_matrix.csv", row.names="gene_id"))
rownames(countData_Gene)<-gsub("\\|.*","",rownames(countData_Gene))
countData_Transcript <- as.matrix(read.csv("transcript_count_matrix.csv", row.names="transcript_id"))
#rownames(countData_Transcript)<-gsub("\\..*","",rownames(countData_Transcript))
colData <- read.csv("Groups.csv", sep=",", row.names=1)
#colData$groups<-as.factor(colData$groups)
#all(rownames(colData) %in% colnames(countData_Gene))
#all(rownames(colData) %in% colnames(countData_Transcript))
countData_Gene <- countData_Gene[, rownames(colData)]
countData_Transcript <- countData_Transcript[, rownames(colData)]
#all(rownames(colData) == colnames(countData_Gene))
#all(rownames(colData) == colnames(countData_Transcript))
countData_Transcript <- countData_Transcript[transcript2genes$transcript_id,]
rownames(countData_Transcript)<-paste0(transcript2genes$gene_name,"_",transcript2genes$transcript_id)
countData_Gene<-countData_Gene[rowSums(countData_Gene>0)>=1,]
countData_Transcript<-countData_Transcript[rowSums(countData_Transcript>0)>=1,]
dds_Gene<-DESeqDataSetFromMatrix(countData = countData_Gene, colData = colData, design = ~ groups)
dds_Gene <- estimateSizeFactors(dds_Gene)
# sizeFactors(dds_Gene)
normalized_counts_Gene <- counts(dds_Gene, normalized=TRUE)
dds_Transcript<-DESeqDataSetFromMatrix(countData = countData_Transcript, colData = colData, design = ~ groups)
dds_Transcript <- estimateSizeFactors(dds_Transcript)
# sizeFactors(dds_Transcript)
normalized_counts_Transcript <- counts(dds_Transcript, normalized=TRUE)
col_groups<-colData$groups
col_groups[col_groups=="Control"]="lightblue"
col_groups[col_groups=="PKO"]="orange"
# Distribution of normalized reads counts in genes/transcripts across all samples
cat("Distribution of normalized reads counts in genes across all samples")
boxplot(log2(normalized_counts_Gene+1),col=col_groups,las=2,ylab='log2(Normalized_Counts+1)')
cat("Distribution of normalized reads counts in transcripts across all samples")
boxplot(log2(normalized_counts_Transcript+1),col=col_groups,las=2,ylab='log2(Normalized_Counts+1)')
pheatmap_color<-data.frame(row.names=rownames(colData), Group=colData$groups)
cat("Heatmap of Whole Gene Expression Profiles")
pheatmap::pheatmap(log2(normalized_counts_Gene+1),scale = "none",angle_col = 45,fontsize = 8,cluster_rows = FALSE,cluster_cols = TRUE,show_rownames = FALSE,annotation_col = pheatmap_color)
cat("Heatmap of Whole Transcript Expression Profiles")
pheatmap::pheatmap(log2(normalized_counts_Transcript+1),scale = "none",angle_col = 45,fontsize = 8,cluster_rows = FALSE,cluster_cols = TRUE,show_rownames = FALSE,annotation_col = pheatmap_color)

```

The clustering suggests outliers in the RNA-seq samples.  

## MDS & PCA  
Multidimensional scaling (MDS) and principal component analysis (PCA) simplify the complexity in high-dimensional data (RNA-seq) while retaining trends and patterns. 

```{r MDS & PCA,fig.height = 8,fig.width = 12,include = TRUE,echo=FALSE,message = FALSE,warning = FALSE,error=TRUE}
# Input the phenotype informartion (treatments, gender, age, etc.) with format as follows:
# ids,groups
# Sample1,Control
# Sample2,test
# ...
# PCA reduces data by geometrically projecting them onto lower dimensions called principal components (PCs), with the goal of finding the best summary of the data using a limited number of PCs.  
# This task can be used to get an impression on the similarity of RNA-seq samples, i.e. to identify subgroups or outliers.  

# The input should be samples (rows) x genes (columns)

t_normalized_counts_Gene<-as.data.frame(t(normalized_counts_Gene))
t_normalized_counts_Transcript<-as.data.frame(t(normalized_counts_Transcript))
# MDS
cat("MDS Graph of individual RNA-seq samples")
mds_Gene <- t_normalized_counts_Gene %>% dist() %>% cmdscale() %>% as_tibble()
colnames(mds_Gene) <- c("Dim.1", "Dim.2")
mds_Gene$groups<-colData$groups
mds_p1<-ggscatter(mds_Gene,title = "MDS by Gene",x = "Dim.1", y = "Dim.2", color = "groups",label = rownames(t_normalized_counts_Gene),palette = "jco",size = 2, ellipse = TRUE,ellipse.type = "convex",repel = TRUE)
mds_Transcript <- t_normalized_counts_Transcript %>% dist() %>% cmdscale() %>% as_tibble()
colnames(mds_Transcript) <- c("Dim.1", "Dim.2")
mds_Transcript$groups<-colData$groups
mds_p2<-ggscatter(mds_Transcript,title = "MDS by Transcript", x = "Dim.1", y = "Dim.2", color = "groups",label = rownames(t_normalized_counts_Transcript),palette = "jco",size = 2, ellipse = TRUE,ellipse.type = "convex",repel = TRUE)
plot_grid(mds_p1,mds_p2,labels = c("A","B"))
# PCA
PCA_Transcript_Results<-prcomp(t_normalized_counts_Transcript)
#cat("Plot the eigenvalues/variances against the number of dimensions.")
#fviz_eig(PCA_Transcript_Results)
cat("PCA Graph of individual RNA-seq samples")
pca_transcript<-fviz_pca_ind(PCA_Transcript_Results,habillage = colData$groups,addEllipses=TRUE, repel = TRUE,ellipse.type = "confidence",palette = c("#00AFBB",  "#FC4E07"),title ="PCA by Transcripts")
PCA_Gene_Results<-prcomp(t_normalized_counts_Gene)
#fviz_eig(PCA_Gene_Results)
pca_gene<-fviz_pca_ind(PCA_Gene_Results,habillage = colData$groups,addEllipses=TRUE, repel = TRUE,ellipse.type = "confidence",palette = c("#00AFBB",  "#FC4E07"),title ="PCA by Genes")
# put the PCA plots by genes/transcripts into one figure
plot_grid(pca_transcript,pca_gene,labels = c("A","B"))

```
PCA identifies outliers from all the samples.  

## Differential Gene Expression Analysis  
After MDS/PCA, the outliers will be removed from the following analysis.  
DESeq2/ballgown are used for analyzing differential expression, which take as input a matrix of read counts mapped to particular genomic features (e.g., genes). 

```{r Differential Gene Expression Analysis,fig.height = 8,fig.width = 12,include = TRUE,echo=FALSE,message = FALSE,warning = FALSE,error=TRUE}
Outlier_samples<-c("COHP_44188","COHP_44190")
pheno_data_filter<-colData %>% filter(!rownames(colData) %in% Outlier_samples)
countData_Gene_filter<-countData_Gene[,!(colnames(countData_Gene) %in% Outlier_samples)]
countData_Transcript_filter<-countData_Transcript[,!(colnames(countData_Transcript) %in% Outlier_samples)]
dds_Gene<-DESeqDataSetFromMatrix(countData = countData_Gene_filter, colData = pheno_data_filter, design = ~ groups)
dds_Gene <- estimateSizeFactors(dds_Gene)
normalized_counts_Gene <- counts(dds_Gene, normalized=TRUE)
dds_Transcript<-DESeqDataSetFromMatrix(countData = countData_Transcript_filter, colData = pheno_data_filter, design = ~ groups)
Average_Norm_Counts_Gene<-data.frame(Average_Norm_Counts_Control=log2(rowMeans(normalized_counts_Gene[,rownames(pheno_data_filter)[pheno_data_filter$groups=="Control"]])+1),Average_Norm_Counts_PKO=log2(rowMeans(normalized_counts_Gene[,rownames(pheno_data_filter)[pheno_data_filter$groups=="PKO"]])+1))
dds_Transcript <- estimateSizeFactors(dds_Transcript)
normalized_counts_Transcript <- counts(dds_Transcript, normalized=TRUE)
Average_Norm_Counts_Transcript<-data.frame(Average_Norm_Counts_Control=log2(rowMeans(normalized_counts_Transcript[,rownames(pheno_data_filter)[pheno_data_filter$groups=="Control"]])+1),Average_Norm_Counts_PKO=log2(rowMeans(normalized_counts_Transcript[,rownames(pheno_data_filter)[pheno_data_filter$groups=="PKO"]])+1))
dds_Gene <- DESeq(dds_Gene)
dds_Transcript <- DESeq(dds_Transcript)
res_Gene <- results(dds_Gene)
res_Transcript <- results(dds_Transcript)
res_Gene_df <- data.frame(res_Gene)
res_Gene_df$Gene_Name<-rownames(res_Gene_df)
res_Gene_df<-merge(res_Gene_df, Average_Norm_Counts_Gene, by=0, all=TRUE)
res_Gene_df<-res_Gene_df[order(res_Gene_df$padj), ]
rownames(res_Gene_df)<-res_Gene_df$Row.names
res_Gene_df<-res_Gene_df[,c(8,9,10,3:7)]
write.table(res_Gene_df,file = "Differentially_Expressed_Genes_RNA_seq.txt",sep = "\t",col.names = TRUE,row.names = FALSE,quote = FALSE)
res_Transcript_df <- data.frame(res_Transcript)
res_Transcript_df$Transcript_Name<-rownames(res_Transcript_df)
res_Transcript_df<-merge(res_Transcript_df, Average_Norm_Counts_Transcript, by=0, all=TRUE)
res_Transcript_df<-res_Transcript_df[order(res_Transcript_df$padj), ]
rownames(res_Transcript_df)<-res_Transcript_df$Row.names
res_Transcript_df<-res_Transcript_df[,c(8,9,10,3:7)]
write.table(res_Transcript_df,file = "Differentially_Expressed_Transcripts_RNA_seq.txt",sep = "\t",col.names = TRUE,row.names = FALSE,quote = FALSE)
cat("Volcano plot of differentially expression analysis in genes")
EnhancedVolcano(res_Gene_df,lab = rownames(res_Gene_df),x = 'log2FoldChange',y = 'pvalue',FCcutoff = 1.1,pointSize = 1,pCutoff = 0.01)
cat("Volcano plot of differentially expression analysis in transcripts")
EnhancedVolcano(res_Transcript_df,lab = rownames(res_Transcript_df),x = 'log2FoldChange',y = 'pvalue',FCcutoff = 1.1,pointSize = 1,pCutoff = 0.01,labSize = 3)

anno<-data.frame(row.names=rownames(pheno_data_filter), Group=pheno_data_filter$groups)
Sig_res_Gene_df<-res_Gene_df %>% filter(pvalue<0.01)
Sig_res_Transcript_df<-res_Transcript_df %>% filter(pvalue<0.01)
cat("Heatmap of Top 50 differentially expressed genes")
pheatmap::pheatmap(normalized_counts_Gene[rownames(Sig_res_Gene_df)[1:50],],scale = "row",annotation_col = anno,angle_col = 45,fontsize = 8)
kbl(Sig_res_Gene_df[1:50,],caption = "Table2. Top 50 differentially expressed genes") %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"),font_size = 11)
cat("Heatmap of Top 50 differentially expressed transcripts")
pheatmap::pheatmap(normalized_counts_Transcript[rownames(Sig_res_Transcript_df)[1:50],],scale = "row",annotation_col = anno,angle_col = 45,fontsize = 8)
kbl(Sig_res_Transcript_df[1:50,],caption = "Table3. Top 50 differentially expressed transcripts") %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"),font_size = 11)

```

## Pathway Enrichment  
Differentially expressed genes detected above need to be further analyzed for their collective functions in order to extract new knowledge. **enrichR** is applied for the analysis.  

```{r Pathway Enrichment,fig.height=10,fig.width=12,include=TRUE,echo=FALSE,message = FALSE,warning = FALSE}
# Canonical pathways are selected
# dbs <- c("KEGG_2019_Human", "Reactome_2016", "BioCarta_2016")
dbs <- c("KEGG_2019_Mouse", "Reactome_2016", "BioCarta_2016")

DEGs_pvalue001 <- res_Gene_df %>% filter(pvalue<0.01)

enriched <- enrichr(DEGs_pvalue001$Gene_Name, dbs)
enriched$KEGG_2019_Mouse <- enriched$KEGG_2019_Mouse %>% filter(P.value<0.05)
enriched$Reactome_2016 <- enriched$Reactome_2016 %>% filter(P.value<0.05)
enriched$BioCarta_2016 <- enriched$BioCarta_2016 %>% filter(P.value<0.05)
cat("Enrichment analysis of gene markers by Enrichr with canonical pathways")
plotEnrich(enriched[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",title = "Enrichment analysis of gene markers in KEGG by Enrichr")
#plotEnrich(enriched[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",title = "Enrichment analysis of gene markers in Reactome by Enrichr")
#plotEnrich(enriched[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",title = "Enrichment analysis of gene markers in BioCarta by Enrichr")
#save.image("RNA_seq_Analysis_Results.RData")
```

## GSEA (Gene Set Enrichment Analysis)  

The package **fgsea** implements an algorithm for fast gene set enrichment analysis.  


```{r GSEA (Gene Set Enrichment Analysis),fig.height=10,fig.width=12,include=TRUE,echo=FALSE,message = FALSE,warning = FALSE}
# Citation:
#[Korotkevich G, Sukhov V, Sergushichev A (2019). “Fast gene set enrichment analysis.” bioRxiv. doi: 10.1101/060012.](http://biorxiv.org/content/early/2016/06/20/060012)  
MSigDB_C2<-msigdbr(species = "mouse", category = "C2")
MSigDB_H<-msigdbr(species = "mouse", category = "H")
gene_sets_canonical<-MSigDB_C2 %>% filter(gs_subcat %in% c("CP:KEGG","CP:BIOCARTA","CP:REACTOME")) %>% dplyr::select(gs_name,gene_symbol)
gene_sets_hallmarks<-MSigDB_H %>% dplyr::select(gs_name,gene_symbol)
Pathway_List <- function(data_frame) {
  unique_pathways<-unique(data_frame$gs_name)
  Pathway_List<-list()
  for(i in 1:length(unique_pathways)){
    Pathway_List[[unique_pathways[i]]]<-data_frame$gene_symbol[data_frame$gs_name==unique_pathways[i]]
    
  }
  return(Pathway_List)
}
Canonical_pathway_list<-Pathway_List(gene_sets_canonical)
Hallmarks_pathway_list<-Pathway_List(gene_sets_hallmarks)
GeneRanks<-res_Gene_df$log2FoldChange
names(GeneRanks)<-res_Gene_df$Gene_Name
#GeneRanks <- sort(GeneRanks, decreasing = TRUE)

cat("GSEA against KEGG, Reactome, ans Biocarta canonical pathways")
fgseaRes_canonical <- fgseaMultilevel(pathways = Canonical_pathway_list,stats = GeneRanks,minSize=10,maxSize=1000,eps = 1e-30)
fgseaRes_canonical<-fgseaRes_canonical[order(pval), ]
fgseaRes_canonical_Sig<-fgseaRes_canonical %>% filter(pval<0.01)
fgseaRes_canonical_Sig$pathway<-gsub("BIOCARTA_|REACTOME_|KEGG_","",fgseaRes_canonical_Sig$pathway)
fgseaRes_canonical_Sig$pathway<-strtrim(fgseaRes_canonical_Sig$pathway,30)
fgseaRes_canonical_Sig$No_of_Significant_Genes<-lengths(fgseaRes_canonical_Sig$leadingEdge)
ggplot(fgseaRes_canonical_Sig, aes(x=NES, y=pathway, size=No_of_Significant_Genes,color=pval)) + geom_point()+scale_size_area(max_size = 10)+scale_colour_gradient(low="blue",high="red")
#topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
#topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=10), pathway]
#topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
#cat(" Table plot for top perturbedpathways:")
#plotGseaTable(Canonical_pathway_list[topPathways], GeneRanks, fgseaRes, 
#              gseaParam=0.5)
#cat(" GSEA plot of top up-regulated pathways:")
#for (i in 1:5) {
#  P1<-plotEnrichment(Canonical_pathway_list[[topPathwaysUp[i]]],GeneRanks) + labs(title=topPathwaysUp[i])
#  print(P1)
#}

#cat(" GSEA plot of down-regulated pathways:")
#for (j in 1:5) {
#  P2<-plotEnrichment(Canonical_pathway_list[[topPathwaysDown[j]]], GeneRanks) + labs(title=topPathwaysDown[j])
#  print(P2)
#}
cat("GSEA against Hallmarks in MSigDB")
fgseaRes_hallmarks <- fgseaMultilevel(pathways = Hallmarks_pathway_list,stats = GeneRanks,minSize=5,maxSize=2000,eps = 1e-60)
fgseaRes_hallmarks<-fgseaRes_hallmarks[order(pval), ]
fgseaRes_hallmarks_Sig<-fgseaRes_hallmarks %>% filter(pval<0.01)
fgseaRes_hallmarks_Sig$pathway<-gsub("HALLMARK_","",fgseaRes_hallmarks_Sig$pathway)
fgseaRes_hallmarks_Sig$pathway<-strtrim(fgseaRes_hallmarks_Sig$pathway,30)
fgseaRes_hallmarks_Sig$No_of_Significant_Genes<-lengths(fgseaRes_hallmarks_Sig$leadingEdge)
ggplot(fgseaRes_hallmarks_Sig, aes(x=NES, y=pathway, size=No_of_Significant_Genes,color=pval)) + geom_point()+scale_size_area(max_size = 10)+scale_colour_gradient(low="blue",high="red")
```

## Save Results  
The Figures and Tables will be automatically saved.   
```{r Save Results,fig.height=8,fig.width=12,include=TRUE,echo=FALSE,message = FALSE,warning = FALSE}
save.image("RNA_seq_Analysis_Results.RData")
```
